python/pyiceberg/avro/resolver.py

```diff
@@ -109,38 +109,46 @@ def resolve(
 
 
 class SchemaResolver(PrimitiveWithPartnerVisitor[IcebergType, Reader]):
-    read_types: Optional[Dict[int, Callable[[Schema], StructProtocol]]]
+    read_types: Dict[int, Type[StructProtocol]]
+    context: List[int]
 
-    def __init__(self, read_types: Optional[Dict[int, Callable[[Schema], StructProtocol]]]):
+    def __init__(self, read_types: Dict[int, Type[StructProtocol]] = EMPTY_DICT) -> None:
         self.read_types = read_types
+        self.context = []
 
     def schema(self, schema: Schema, expected_schema: Optional[IcebergType], result: Reader) -> Reader:
         return result
 
+    def before_field(self, field: NestedField, field_partner: Optional[NestedField]) -> None:
+        self.context.append(field.field_id)
+
+    def after_field(self, field: NestedField, field_partner: Optional[NestedField]) -> None:
+        self.context.pop()
+
     def struct(self, struct: StructType, expected_struct: Optional[IcebergType], field_readers: List[Reader]) -> Reader:
+        # -1 indicates the struct root
+        read_struct_id = self.context[-1] if len(self.context) > 0 else -1
+        struct_callable = self.read_types.get(read_struct_id, Record)
```

Comment:
I just went to refer back to my later changes, but I see that [20b07d35](https://github.com/apache/iceberg/pull/6506/commits/20b07d355f2238580273b83212a6f25ea0d62f21) is no longer there. I pushed the changes here: https://github.com/apache/iceberg/commit/feef2b6a5d7d33f0818e9ed1e36b31c5442e050d

This PR changes `read_types` by passing in just the type, but this needs to pass the read schema to the type. Without the read schema, the positions used by `set` won't be correct. For example, if the file contains int columns a, b, and c, but the read schema requested b and d, this will produce `((None, IntReader), (0, IntReader), (None, IntReader), (1, NoneReader))`. Then the struct instance will get calls `set(0, reader.read(decoder))` and `set(1, noneReader.read(decoder))`. So the instance needs to know what positions 0 and 1 are.

There are two solutions to the problem. The first is to pass the read schema to the class so it can map the fields, which is what I did in my PR. That's how `IndexedRecord` works in the JVM Avro implementation.

The second solution is to use IDs rather than positions. If we did that, then some `setById` method on the struct instance would always know what to do. This would require a different `StructProtocol` or an extension to it, and I think we would still want to have a way to pass the struct type to generic `Record` instances... so I think for now it makes sense to go with the position approach and pass the schema to the struct instance when it is created.

Reply:
I also spend quite some CPU cycles on this, and I think it will go alright since we already solve this when building the read-tree. We have to file schema from the file, and the read schema from `manifest.py`. It is very important that we align the Pydantic classes and the schema, if these aren't aligned, it will be assigned to the incorrect field.

I've added a test to validate this:
```python
def test_column_assignment() -> None:
    int_schema = {
        "type": "record",
        "name": "ints",
        "fields": [
            {"name": "a", "type": "int", "field-id": 1},
            {"name": "b", "type": "int", "field-id": 2},
            {"name": "c", "type": "int", "field-id": 3},
        ],
    }

    from fastavro import parse_schema, writer

    parsed_schema = parse_schema(int_schema)

    int_records = [
        {
            "a": 1,
            "b": 2,
            "c": 3,
        }
    ]

    with TemporaryDirectory() as tmpdir:
        tmp_avro_file = tmpdir + "/manifest.avro"
        with open(tmp_avro_file, "wb") as out:
            writer(out, parsed_schema, int_records)

        class Ints(PydanticStruct):
            c: int = Field()
            d: Optional[int] = Field()

        MANIFEST_ENTRY_SCHEMA = Schema(
            NestedField(3, "c", IntegerType(), required=True),
            NestedField(4, "d", IntegerType(), required=False),
        )

        with AvroFile[Ints](PyArrowFileIO().new_input(tmp_avro_file), MANIFEST_ENTRY_SCHEMA, {-1: Ints}) as reader:
            records = list(reader)

    assert records == [Ints(c=3, d=None)]
```

I've confirmed that `__set__` is called twice, one time with the `c` column, and then a `NoneReader`.

Reply:
@Fokko, I think that the problem with this test is that the class `Ints` matches `MANIFEST_ENTRY_SCHEMA`. To test what I'm talking about, `Ints` should match the file schema. That corresponds to the case where we have `ManifestFile` that has all of the manifest file fields, but read just a subset of those fields when projecting.

Reply:
Ah, then we might have a miscommunication somewhere. The whole implementation is based on the idea of aligning the struct and the read schema (by position). This makes it so elegant since the position of the struct and read-schema align. If we should match the fields with the read schema (by name), then we're back to the original Pydantic implementation.

Reply:
I think that the implementation is correct for the case you describe and we do want to support that. What I think we need to support for the metadata structures is different because we have the write schema, read schema, and record schema. When the read schema and the record schema don't match, we let the record handle it.

We don't want to get rid of the read schema because that removes the ability to project columns of the record schema. We can't get rid of the record schema because the record implicitly has one. We just need to account for cases where they differ.

It's also better to pass the read schema because we can make generic records more friendly.

Reply:
It looks like we can use `inspect.getfullargspec(class.__init__)` to inspect the init arguments. So we could do something like this:

```python
    t = self.read_types.get(field_id)
    if t:
        spec = inspect.getfullargspec(t.__init__)
        if len(spec.args) > 1 and 'schema' in spec.args:
            create_func = lambda schema: t(**{'schema': read_schema})
        else if spec.varkw:
            create_func = lambda schema: t(**{'schema': read_schema})
        else:
            create_func = t
```

That would allow us to keep passing types (which I like) but also pass the schema if it is supported either by kwargs

Reply:
Got it. Currently, there is no record schema for the typed classes (`ManifestEntry`, `ManifestFile`, etc). I would be hesitant to add yet another schema there. Are there any cases where we want to have the Record schema and the Read schema differ? Ideally, you want to shape the read schema to the data that you need, so you don't read anything that's not being used.

> It's also better to pass the read schema because we can make generic records more friendly.

This is something that we do for the `Record` class:

```python
    def read(self, decoder: BinaryDecoder) -> Any:
        if issubclass(self.create_struct, Record):
            struct = self.create_struct(length=len(self.field_readers), fields=self.fields)
        elif issubclass(self.create_struct, PydanticStruct):
            struct = self.create_struct.construct()
        else:
            raise ValueError(f"Expected a subclass of PydanticStruct, got: {self.create_struct}")
```

And then we have the attributes available:
```python
def test_named_record() -> None:
    r = Record(fields=(NestedField(0, "id", IntegerType()), NestedField(1, "name", StringType())))

    assert r.id is None
    assert r.name is None

    r[0] = 123
    r[1] = "abc"

    assert r[0] == 123
    assert r[1] == "abc"

    assert r.id == 123
    assert r.name == "abc"
```

> It looks like we can use inspect.getfullargspec(class.__init__) to inspect the init arguments. So we could do something like this:

I'm not a big fan of doing inspections, similar to Java, it is quite expensive. And it looks like there is a [performance issue](https://bugs.python.org/issue37010) in `getfullargspec`. What do you think of passing in the fields as we have currently?

Reply:
There is a need to be able to read a subset of fields when reading a manifest list or a manifest file. Right now, we don't use the lower or upper bounds, for example, but we still read them into memory rather than skipping them. There are also other use cases. A generic record, for example, should be able to use the field names from the read schema.

To do those things, this PR needs to support passing the read schema to the instance that is instantiated. I don't really care how we do it. It can be by inspecting the constructor, assuming that the constructor accepts a schema, checking for a `set_read_schema` method, or some other idea. But it needs to be done as part of this PR.